### PR TITLE
Handle wrapping PlaceholderResolverProvider in ConfigServerHostedService

### PR DESCRIPTION
**Pull request by TimHess**
## Description

Handle the situation where `ConfigServerHostedService` is given an `IConfigurationRoot` where the `PlaceholderResolverProvider` wraps the `ConfigServerConfigurationProvider`

Addresses #879, needs backport to `release/3.2`

## Quality checklist

- [x] Your code complies with our [Coding Style](https://github.com/SteeltoeOSS/.github/blob/main/contributing-docs/contributing-code-style.md).
- [x] You've updated unit and/or integration tests for your change, where applicable.
- [x] You've updated documentation for your change, where applicable.
      If your change affects other repositories, such as [Documentation](https://github.com/SteeltoeOSS/Documentation), [Samples](https://github.com/SteeltoeOSS/Samples) and/or [MainSite](https://github.com/SteeltoeOSS/MainSite), add linked PRs here.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.
- [x] You've added required license files and/or file headers (explaining where the code came from with proper attribution), where code is copied from StackOverflow, a blog, or OSS.

### CODE REVIEW

File: src/Configuration/src/ConfigServerBase/ConfigServerHostedService.cs [hunk 1]
```diff
@@ -30,7 +31,17 @@ public ConfigServerHostedService(IConfigurationRoot configuration, ILoggerFactor
             throw new ArgumentNullException(nameof(configuration));
         }
 
-        _configuration = configuration.Providers.First(provider => provider is ConfigServerConfigurationProvider) as ConfigServerConfigurationProvider;
+        if (configuration.Providers.Count() == 1 && configuration.Providers.First() is PlaceholderResolverProvider resolverProvider)
+        {
+            _configuration =
+                resolverProvider.Providers.First(provider => provider is ConfigServerConfigurationProvider) as
+                    ConfigServerConfigurationProvider;
```
Comment (bart-vmware) [inline src/Configuration/src/ConfigServerBase/ConfigServerHostedService.cs:38]: Could it happen that `ConfigServerConfigurationProvider` is not the first one? The following would be more reliable:
```c#
_configuration = resolverProvider.Providers.OfType<ConfigServerConfigurationProvider>().First();
```

File: src/Configuration/Configuration.sln [hunk 2]
```diff
@@ -98,7 +98,15 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Steeltoe.Management.Abstrac
...
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Steeltoe.Common.Utils", "..\Common\src\Common.Utils\Steeltoe.Common.Utils.csproj", "{F3D90FC9-FAB5-49EB-9A4D-BBF28F8116AA}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Steeltoe.Discovery.ClientBase", "..\Discovery\src\ClientBase\Steeltoe.Discovery.ClientBase.csproj", "{7F1D9558-4908-4BB4-B39E-F507432ACF99}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Steeltoe.Extensions.Logging.DynamicLogger", "..\Logging\src\DynamicLogger\Steeltoe.Extensions.Logging.DynamicLogger.csproj", "{3A265CB8-530E-4FC8-931D-01DE4E7F3F61}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Steeltoe.Connector.ConnectorBase", "..\Connectors\src\ConnectorBase\Steeltoe.Connector.ConnectorBase.csproj", "{36F12A1C-00C2-4049-8A0C-7595C6185F01}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Steeltoe.Extensions.Logging.Abstractions", "..\Logging\src\Abstractions\Steeltoe.Extensions.Logging.Abstractions.csproj", "{C86FB6E5-3024-4028-9BC0-602AEBA65E5B}"
```
[the remark this answers is not in the document]
Comment (TimHess) [reply src/Configuration/Configuration.sln:109]: I believe this was intentional as I'm pretty sure the dependencies didn't want to compile without the additional references. I can double check that there aren't any extras here, but I think they _should_ have been there before

File: src/Configuration/test/ConfigServerBase.Test/ConfigServerHostedServiceTest.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    public async Task ServiceConstructsAndOperatesWithConfigurationManager()
+    {
+        var configurationManager = new ConfigurationManager();
+        configurationManager.AddInMemoryCollection(TestHelpers._fastTestsConfiguration);
+        configurationManager.AddConfigServer();
+        var service = new ConfigServerHostedService(configurationManager, null);
+
+        await service.StartAsync(default);
+        await service.StopAsync(default);
+        Assert.True(true, "Service constructed, started and stopped without exception");
```
[the remark this answers is not in the document]
Comment (TimHess) [reply src/Configuration/test/ConfigServerBase.Test/ConfigServerHostedServiceTest.cs:36]: Fair point, I rushed a bit with these and was thinking it was close enough

File: src/Configuration/src/ConfigServerBase/ConfigServerHostedService.cs [hunk 4]
```diff
@@ -30,7 +31,17 @@ public ConfigServerHostedService(IConfigurationRoot configuration, ILoggerFactor
             throw new ArgumentNullException(nameof(configuration));
         }
 
-        _configuration = configuration.Providers.First(provider => provider is ConfigServerConfigurationProvider) as ConfigServerConfigurationProvider;
+        if (configuration.Providers.Count() == 1 && configuration.Providers.First() is PlaceholderResolverProvider resolverProvider)
```
[the remark this answers is not in the document]
Comment (bart-vmware) [reply src/Configuration/src/ConfigServerBase/ConfigServerHostedService.cs:34]: Ok, makes sense.

File: src/Configuration/src/ConfigServerBase/ConfigServerHostedService.cs [hunk 5]
```diff
@@ -30,7 +31,17 @@ public ConfigServerHostedService(IConfigurationRoot configuration, ILoggerFactor
...
+        {
+            _configuration =
+                resolverProvider.Providers.First(provider => provider is ConfigServerConfigurationProvider) as
+                    ConfigServerConfigurationProvider;
+        }
+        else
+        {
+            _configuration = configuration.Providers.First(provider => provider is ConfigServerConfigurationProvider) as ConfigServerConfigurationProvider;
```
Comment (TimHess) [inline src/Configuration/src/ConfigServerBase/ConfigServerHostedService.cs:42]: Application of @bart-vmware suggestion
```suggestion
            _configuration = resolverProvider.Providers.OfType<ConfigServerConfigurationProvider>().First();
        }
        else
        {
            _configuration = configuration.Providers.OfType<ConfigServerConfigurationProvider>().First();
```

File: src/Configuration/test/ConfigServerBase.Test/ConfigServerHostedServiceTest.cs (hunk 3 above)
Comment (TimHess) [inline src/Configuration/test/ConfigServerBase.Test/ConfigServerHostedServiceTest.cs:36]: @bart-vmware this look better?
```suggestion
        var startAction = async () => await service.StartAsync(default);
        var stopAction = async () => await service.StopAsync(default);

        await startAction.Should().NotThrowAsync("ConfigServerHostedService should start");
        await stopAction.Should().NotThrowAsync("ConfigServerHostedService should be stoppable");
```

File: src/Configuration/test/ConfigServerBase.Test/ConfigServerHostedServiceTest.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+            new List<IConfigurationProvider>
+            {
+                new MemoryConfigurationProvider(_fastTests),
+                new ConfigServerConfigurationProvider()
+            });
+        var service = new ConfigServerHostedService(configurationRoot, null);
+
+        await service.StartAsync(default);
+        await service.StopAsync(default);
+        Assert.True(true, "Service constructed, started and stopped without exception");
```
Comment (TimHess) [inline src/Configuration/test/ConfigServerBase.Test/ConfigServerHostedServiceTest.cs:53]:
```suggestion
        var startAction = async () => await service.StartAsync(default);
        var stopAction = async () => await service.StopAsync(default);

        await startAction.Should().NotThrowAsync("ConfigServerHostedService should start");
        await stopAction.Should().NotThrowAsync("ConfigServerHostedService should be stoppable");
```

File: src/Configuration/test/ConfigServerBase.Test/ConfigServerHostedServiceTest.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            {
+                new MemoryConfigurationProvider(_fastTests),
+                new ConfigServerConfigurationProvider()
+            });
+        var configRoot = new ConfigurationRoot(new List<IConfigurationProvider> { placeholder });
+        var service = new ConfigServerHostedService(configRoot, null);
+
+        await service.StartAsync(default);
+        await service.StopAsync(default);
+        Assert.True(true, "Service constructed, started and stopped without exception");
```
Comment (TimHess) [inline src/Configuration/test/ConfigServerBase.Test/ConfigServerHostedServiceTest.cs:70]:
```suggestion
        var startAction = async () => await service.StartAsync(default);
        var stopAction = async () => await service.StopAsync(default);

        await startAction.Should().NotThrowAsync("ConfigServerHostedService should start");
        await stopAction.Should().NotThrowAsync("ConfigServerHostedService should be stoppable");
```

File: src/Configuration/src/ConfigServerBase/ConfigServerHostedService.cs (hunk 5 above)
Comment (TimHess) [reply src/Configuration/src/ConfigServerBase/ConfigServerHostedService.cs:42]: image

File: src/Configuration/test/ConfigServerBase.Test/ConfigServerHostedServiceTest.cs (hunk 3 above)
Comment (TimHess) [reply src/Configuration/test/ConfigServerBase.Test/ConfigServerHostedServiceTest.cs:36]: updated, better now?

Comment (bart-vmware) [reply src/Configuration/test/ConfigServerBase.Test/ConfigServerHostedServiceTest.cs:36]: This is basically point 7 at https://ddoomen.medium.com/dennis-17-laws-of-test-driven-development-74122e7582c:
> Only have one scenario per test. So doing something, asserting the outcome, then doing something else and then again asserting the outcome should be avoided. It makes the test harder to read, and will make it skip the second or third scenario if the first one fails.

File: src/Configuration/test/ConfigServerBase.Test/ConfigServerHostedServiceTest.cs (hunk 6 above)
Comment (bart-vmware) [reply src/Configuration/test/ConfigServerBase.Test/ConfigServerHostedServiceTest.cs:53]: Yes, I'm okay with that.

**Review by bart-vmware — APPROVED** [review]

**Review by bart-vmware — APPROVED** [review]
LGTM

**Pull request merged**